hmpps-prisoner-search-indexer/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/indexer/services/PrisonerDifferenceService.kt

```diff
@@ -0,0 +1,213 @@
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.TelemetryEvents.PRISONER_CREATED
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.TelemetryEvents.PRISONER_UPDATED
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.TelemetryEvents.PRISONER_UPDATED_DB_NO_CHANGE
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.TelemetryEvents.PRISONER_UPDATED_NO_DIFFERENCES
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.TelemetryEvents.PRISONER_UPDATED_OS_NO_CHANGE
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.trackEvent
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.config.trackPrisonerEvent
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.repository.PrisonerHashRepository
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.services.dto.nomis.OffenderBooking
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.services.events.AlertsUpdatedEventService
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.services.events.HmppsDomainEventEmitter
+import uk.gov.justice.digital.hmpps.prisonersearch.indexer.services.events.PrisonerMovementsEventService
+import java.time.Instant
+import kotlin.reflect.full.findAnnotations
+
+@Service
+class PrisonerDifferenceService(
+  private val telemetryClient: TelemetryClient,
+  private val domainEventEmitter: HmppsDomainEventEmitter,
+  private val diffProperties: DiffProperties,
+  private val prisonerHashRepository: PrisonerHashRepository,
+  private val objectMapper: ObjectMapper,
+  private val prisonerMovementsEventService: PrisonerMovementsEventService,
+  private val alertsUpdatedEventService: AlertsUpdatedEventService,
+) {
+  private companion object {
+    private val log: Logger = LoggerFactory.getLogger(this::class.java)
+    private val exemptedMethods = listOf("diff", "equals", "toString", "hashCode")
+  }
+
+  internal val propertiesByDiffCategory: Map<DiffCategory, List<String>> =
+    Prisoner::class.members
+      .filter { property -> property.findAnnotations<DiffableProperty>().isNotEmpty() }
+      .groupBy { property -> property.findAnnotations<DiffableProperty>().first().type }
+      .mapValues { propertiesByDiffCategory -> propertiesByDiffCategory.value.map { property -> property.name } }
+
+  internal val diffCategoriesByProperty: Map<String, DiffCategory> =
+    Prisoner::class.members
+      .filter { property -> property.findAnnotations<DiffableProperty>().isNotEmpty() }
+      .associate { property -> property.name to property.findAnnotations<DiffableProperty>().first().type }
+
+  @Transactional
+  fun handleDifferences(previousPrisonerSnapshot: Prisoner?, offenderBooking: OffenderBooking, prisoner: Prisoner) {
+    val newHash = prisoner.hash()
+    if (previousPrisonerSnapshot?.hash() != newHash) {
+      if (updateDbHash(offenderBooking.offenderNo, newHash)) {
```

Comment:
split out into multiple if blocks so can tell the difference between open search record changing and database record changing